nerfstudio/model_components/renderers.py

```diff
@@ -101,26 +101,96 @@ def combine_rgb(
+    ) -> Union[Float[Tensor, "3"], Float[Tensor, "*bs 3"]]:
+        """Returns the RGB background color for a specified background color.
+
+        Args:
+            rgb: RGB for each sample.
+            background_color: The background color specification.
+
+        Returns:
+            Background color as RGB.
+        """
         if BACKGROUND_COLOR_OVERRIDE is not None:
             background_color = BACKGROUND_COLOR_OVERRIDE
-        if background_color == "last_sample":
+        elif background_color == "last_sample":
             background_color = rgb[..., -1, :]
-        if background_color == "random":
-            background_color = torch.rand_like(comp_rgb).to(rgb.device)
-        if isinstance(background_color, str) and background_color in colors.COLORS_DICT:
+        elif background_color == "random":
+            background_color = torch.rand_like(rgb[..., -1, :]).to(rgb.device)
+        elif isinstance(background_color, str) and background_color in colors.COLORS_DICT:
             background_color = colors.COLORS_DICT[background_color].to(rgb.device)
+        assert isinstance(background_color, Tensor)
+        return background_color
 
-        assert isinstance(background_color, torch.Tensor)
-        comp_rgb = comp_rgb + background_color.to(weights.device) * (1.0 - accumulated_weight)
+    @classmethod
+    def blend_background(
+        cls,
+        rgb: Tensor,
+        opacity: Tensor,
+        background_color: Float[Tensor, "3"],
+    ) -> Float[Tensor, "*bs 3"]:
+        """Blends the background color into the image.
 
-        return comp_rgb
+        Args:
+            rgb: RGB per pixel.
+            opacity: Alpha opacity per pixel.
+            background_color: Background color.
+
+        Returns:
+            Blended RGB.
+        """
+        if len(opacity.shape) < len(rgb.shape):
+            opacity = opacity.unsqueeze(-1)
+        blended_rgb = rgb * opacity + background_color.to(rgb.device) * (1 - opacity)
+        return blended_rgb
+
+    def blend_background_for_metric_computation(
+        self,
+        gt_image: Tensor,
+        predicted_rgb: Tensor,
+        predicted_opacity: Tensor,
+        background_color: Float[Tensor, "3"] = colors.BLACK,
+    ) -> Tuple[Tensor, Tensor]:
+        """Blends a background color into the ground truth and predicted image for
+        metric computation if the current background color is "random".
+
+        Args:
+            gt_image: The ground truth image.
+            predicted_rgb: The predicted RGB values.
+            predicted_opacity: The predicted opacity/ accumulation.
+            background_color: The background color to use for blending. This should be
+                a single color to use for all pixels.
+
+        Returns:
+            A tuple of the predicted and ground truth RGB values.
+        """
+        if gt_image.shape[-1] == 4 and self.background_color == "random":
+            predicted_rgb = self.blend_background(
```

Comment:
This is not correct, right? For predicted RGB the background colour should be just "added" not "blended". This would give different (incorrect results).

Reply:
Can we have two methods? "blend_background" and "add_background"? "blend_background" would be called for the GT images and 'add_background" for predictions?

Reply:
Accumulation is not binary, right?

Reply:
> Accumulation is not binary, right?

It should be once it converges to the desired alpha. So I don't think it matters for metric computation. Made a comment about this [here](https://github.com/nerfstudio-project/nerfstudio/pull/2165#discussion_r1254305734).

Reply:
> Also, prediction blending should happen regardless of whether the gt has transparency. In fact, non-transparent bg recovers the previous behaviour.

For your changes, this is correct. The changes currently on this PR still due the previous blending before this function is called.